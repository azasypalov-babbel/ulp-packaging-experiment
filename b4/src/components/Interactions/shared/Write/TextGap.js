import React, { forwardRef, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import useGapWidth from './useGapWidth';
import { isiOS } from '../../../../lib/features';

const StyledSpan = styled.span`
    width: ${({ gapWidth }) => gapWidth};
`;

const TextGap = forwardRef(({ text, targetText, alternativeText, ...props }, upperRef) => {
  const ref = useRef();
     // iOS caret increases the required gap width
  const adjustForCaret = isiOS() && !text?.trim();
  const gapWidth = useGapWidth({ ref: upperRef || ref, targetText, adjustForCaret });

  return (
    <StyledSpan
      ref={upperRef || ref}
      role="textbox"
      data-solution={targetText}
      data-solution-alternative={alternativeText}
      gapWidth={gapWidth}
      {...props}
    >
      {text || ' '}
    </StyledSpan>
  );
});

TextGap.displayName = 'TextGap';

TextGap.propTypes = {
  text: PropTypes.string,
  targetText: PropTypes.string.isRequired,
  alternativeText: PropTypes.string
};

export default TextGap;
