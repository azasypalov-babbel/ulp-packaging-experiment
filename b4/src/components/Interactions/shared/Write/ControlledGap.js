import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';

import StyledGap from './StyledGap';
import TextGap from './TextGap';

import { targetChoice, attemptType } from '../scene/scenePropTypes';

import { getAlternativeSolution } from './helpers';
import { RESULTS } from '../../../../lib/matchingUtils/evaluate';
import { useSelector } from 'react-redux';

const stripHtml = (s) => s ? s.replace(/(<([^>]+)>)/ig, '') : '';
const stopPropagation = (event) => event.stopPropagation();

const refocus = (event) => {
  // event.persist() should be called when using React synthetic events inside an asynchronous callback function
  event.persist()
  // Firefox Fix: a timer to make sure that your code runs in an event loop separate from the one triggering the "blur" event. https://bugzilla.mozilla.org/show_bug.cgi?id=120995
  setTimeout(() => {
    if (event.relatedTarget === null) {
      event.target.focus({ preventScroll: true })
    }
  }, 1);
};

const ControlledGap = ({
  active,
  latestText,
  showFeedback,
  targetChoices,
  attempt,
  onKeyDown
}) => {
  const ref = useRef();
  const isZendeskWidgetOpen = useSelector(state => state.zendeskWidget.isOpen);
  useEffect(() => {
    if (active && !showFeedback && !isZendeskWidgetOpen) {
      ref.current?.focus({ preventScroll: true });
    }
  }, [ref, active, showFeedback, isZendeskWidgetOpen]);

  const targetTexts = targetChoices
    .filter((choice) => choice.correct)
    .map((choice) => choice.sentence);

  // PH should disregard multiple solutions
  const targetText = stripHtml(targetTexts[0]);
  const inputTextGap = stripHtml(attempt.inputText);
  const textGap = stripHtml(attempt.text);

  // alternative solution
  const alternativeText = getAlternativeSolution(targetTexts, targetText);

  if (!active) {
    return (
      <StyledGap active={false}>
        <TextGap
          text={attempt.number ? textGap : 'Â '}
          targetText={attempt.number ? textGap : targetText}
          alternativeText={alternativeText}
        />
      </StyledGap>
    );
  }

  if (showFeedback && attempt.solved) {
    const hasTypoFeedback = attempt.feedbackType === RESULTS.TYPO;
    return (
      <StyledGap
        active={true}
        solved={true}
        typo={hasTypoFeedback}
      >
        <TextGap
          text={hasTypoFeedback ? inputTextGap : textGap}
          targetText={hasTypoFeedback ? inputTextGap : textGap}
          alternativeText={alternativeText}
        />
      </StyledGap>
    );
  }

  return (
    <StyledGap
      active={true}
      onClick={stopPropagation}
      mistaken={!attempt.solved && showFeedback}
    >
      <TextGap
        ref={ref}
        tabIndex={0}
        onBlur={!isZendeskWidgetOpen ? refocus : undefined}
        data-selector="active-gap"
        data-focus-on-key="Shift+Escape"
        onKeyDown={onKeyDown}
        text={latestText}
        targetText={targetText}
        alternativeText={alternativeText}
      />
    </StyledGap>
  );
};

ControlledGap.displayName = 'ControlledGap';

ControlledGap.propTypes = {
  onKeyDown: PropTypes.func,
  showFeedback: PropTypes.bool,
  active: PropTypes.bool.isRequired,
  latestText: PropTypes.string,
  targetChoices: PropTypes.arrayOf(targetChoice).isRequired,
  attempt: attemptType
};

export default ControlledGap;
