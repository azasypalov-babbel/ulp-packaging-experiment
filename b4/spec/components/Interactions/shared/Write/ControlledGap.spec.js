import React from 'react';
import { shallow } from 'enzyme';
import * as redux from 'react-redux'
import ControlledGap from '../../../../../src/components/Interactions/shared/Write/ControlledGap';
import TextGap from '../../../../../src/components/Interactions/shared/Write/TextGap';

const spy = jest.spyOn(redux, 'useSelector')
spy.mockReturnValue({ zendeskWidget: { isOpen: false } })

const defaultProps = {
  active: true,
  latestText: 'abc',
  targetChoices: [
    { sentence: 'answer', correct: true, oddSolution: false },
    { sentence: 'alternative answer', correct: true, oddSolution: false }
  ],
  attempt: {
    text: 'answer',
    solved: false,
    feedbackType: 'CORRECT',
    inputText: 'abc',
    mistaken: false,
    pending: false,
    number: 1
  },
  onCheat: () => {},
  feedbackDismissed: false,
  setFeedbackDismissed: jest.fn()
};

describe('Controlled Gap', () => {
  beforeEach(() => {
  });

  describe('renders', () => {
    test('html when active', () => {
      const component = shallow(<ControlledGap {...defaultProps} />);
      expect(component).toMatchSnapshot();
    });

    test('html when not active', () => {
      const component = shallow(<ControlledGap {...defaultProps} active={false} />);
      expect(component).toMatchSnapshot();
    });

    test('a TextGap with props', () => {
      const component = shallow(<ControlledGap {...defaultProps} />);
      const textGap = component.find(TextGap);

      expect(textGap.prop('text')).toBe(defaultProps.latestText);
      expect(textGap.prop('targetText')).toBe(defaultProps.targetChoices[0].sentence);
      expect(textGap.prop('alternativeText')).toBe(defaultProps.targetChoices[1].sentence);
    });
  });
});
