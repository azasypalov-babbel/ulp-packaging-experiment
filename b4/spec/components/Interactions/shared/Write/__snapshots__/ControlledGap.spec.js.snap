// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controlled Gap renders html when active 1`] = `
<styled.span
  active={true}
  onClick={[Function]}
>
  <TextGap
    alternativeText="alternative answer"
    data-focus-on-key="Shift+Escape"
    data-selector="active-gap"
    tabIndex={0}
    targetText="answer"
    text="abc"
  />
</styled.span>
`;

exports[`Controlled Gap renders html when not active 1`] = `
<styled.span
  active={false}
>
  <TextGap
    alternativeText="alternative answer"
    targetText="answer"
    text="answer"
  />
</styled.span>
`;
